services/horizon/internal/ingest/processors/asset_stats_processor.go

```
@@ -230,5 +282,92 @@ func (p *AssetStatsProcessor) Commit(ctx context.Context) error {
 		}
 	}
 
+	return p.updateContractIDs(ctx, contractToAsset)
+}
+
+func (p *AssetStatsProcessor) updateContractIDs(ctx context.Context, contractToAsset map[[32]byte]*xdr.Asset) error {
+	for contractID, asset := range contractToAsset {
+		if err := p.updateContractID(ctx, contractID, asset); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *AssetStatsProcessor) updateContractID(ctx context.Context, contractID [32]byte, asset *xdr.Asset) error {
+	var rowsAffected int64
+	if asset == nil {
+		stat, err := p.assetStatsQ.GetAssetStatByContract(ctx, contractID)
+		if err == sql.ErrNoRows {
+			return ingest.NewStateError(errors.Errorf(
+				"row for asset with contract %s is missing",
+				hex.EncodeToString(contractID[:]),
+			))
+		}
+		if err != nil {
+			return errors.Wrap(err, "could not find asset stat by contract id")
+		}
+
+		if stat.Accounts.IsZero() {
+			rowsAffected, err = p.assetStatsQ.RemoveAssetStat(ctx,
+				stat.AssetType,
+				stat.AssetCode,
+				stat.AssetIssuer,
+			)
+			if err != nil {
+				return errors.Wrap(err, "could not remove asset stat")
+			}
+		} else {
+			stat.ContractID = nil
+			rowsAffected, err = p.assetStatsQ.UpdateAssetStat(ctx, stat)
+			if err != nil {
+				return errors.Wrap(err, "could not update asset stat")
+			}
+		}
+	} else {
+		var assetType xdr.AssetType
+		var assetCode, assetIssuer string
+		asset.MustExtract(&assetType, &assetCode, &assetIssuer)
```

does it make sense to panic or is it better to return an error like the other cases?